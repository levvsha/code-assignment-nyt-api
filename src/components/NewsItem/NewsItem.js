import './NewsItem.styl';

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _debounce from 'lodash.debounce';

export default class NewsItem extends Component {
  static propTypes = {
    headline: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
    id: PropTypes.string.isRequired,
    snippet: PropTypes.string,
    image: PropTypes.string,
    isActive: PropTypes.bool,
    publicationDate: PropTypes.string,
    link: PropTypes.string
  };

  state = {
    expandedImageHeight: null
  }

  constructor() {
    super();

    this.debouncedHandleResize = _debounce(this.handleResize, 350);
  }

  componentDidMount() {
    window.addEventListener('resize', this.debouncedHandleResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.debouncedHandleResize);
  }

  handleResize = () => {
    const imageHeight = this.image && this.image.getBoundingClientRect().height;

    this.setState({ expandedImageHeight: imageHeight });
  }

  handleClick = () => {
    if (!this.props.isActive) {
      const imageHeight = this.image && this.image.getBoundingClientRect().height;

      this.setState({ expandedImageHeight: imageHeight });
    }

    this.props.onClick(this.props.id);
  }

  render() {
    const {
      headline,
      snippet,
      image,
      isActive,
      publicationDate,
      link,
    } = this.props;

    const { expandedImageHeight } = this.state;

    return (
      <div
        className={`c-news-item${ isActive ? ' active' : '' }`}
        onClick={this.handleClick}
      >
        <div
          className="item-media"
          style={isActive && expandedImageHeight ? { minHeight: `${ expandedImageHeight }px` } : null}
        >
          <div className="image-wrapper">
            <img ref={image => (this.image = image)} src={image} alt="" />
          </div>
        </div>
        <div className="item-body">
          <h3 className="item-title">
            {headline}
          </h3>

          <div className="item-description">
            <p className="item-published">
              Publisherd: {publicationDate}
            </p>
            <p className="item-snippet">
              {snippet}
            </p>
            <p className="read-more">
              <a
                target="_blank"
                href={link}
              >
                Read more
              </a>
            </p>
          </div>
        </div>
      </div>
    );
  }
}
