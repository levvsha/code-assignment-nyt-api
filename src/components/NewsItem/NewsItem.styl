.c-news-item {
  display: flex
  align-items: start
  background-color: #fff
  padding: 15px
  margin: 24px 15px 15px
  border-radius: 10px
  cursor: pointer
  box-shadow: 0 0 30px rgba(0, 0, 0, 0.1)
  getTransition(background-color)

  &.active {
    cursor: default
    background-color: $grayBlue

    .item-description {
      opacity: 1
      max-height: 200px
    }
  }

  .item-media {
    font-size: 0
    width: 350px
    min-width: 350px
    overflow: hidden
    max-height: 100%
    position: relative
    border-radius: 10px
    min-height: 60px
    getTransition(min-height)
  }

  .image-wrapper {
    position: absolute
    font-size: 0
    left: 0
    width: 100%
    border-radius: 10px
    overflow: hidden
    top: 50%
    transform: translateY(-50%)
  }

  .item-published {
    font-size: 12px
    color: #d6c7ef
    text-align: right
    margin: 7px 0
  }

  .read-more {
    margin-bottom: 0

    a {
      font-size: 14px
      color: #d6c7ef
      cursor: pointer


      &:hover {
        text-decoration: none
      }
    }
  }

  .item-media img {
    width: 100%
  }

  .item-body {
    padding-left: 15px
    min-height: 60px
  }

  .item-title {
    font-size: 21px
    font-weight: 600
    line-height: 1.30em
    margin: 0
    padding: 0
  }

  .item-description {
    opacity: 0
    max-height: 0
    overflow: hidden
    transition: max-height 0.3s ease-in-out 0s, opacity 0.3s ease-in-out 0s
    margin: 0
    padding: 0
}

  @media screen and (max-width: 836px) {
    .item-media {
      width: 225px
      min-width: 225px
    }
  }

  @media screen and (max-width: 680px) {
    flex-direction: column
    min-height: auto

    .item-media {
      min-height: 90px
    }

    .item-body {
      margin-top: 10px
    }

    .item-media {
      width: 100%
    }
  }
}
