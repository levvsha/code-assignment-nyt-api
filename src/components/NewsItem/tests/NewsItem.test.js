import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import NewsItem from '../';

describe('<NewsItem />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('NewsItem renders correctly', () => {
    const tree = renderer.create(
      <NewsItem
        headline="headline"
        onClick={() => {}}
        id="id"
        snippet="snippet"
        image="url/to/image.jpg"
        isActive={false}
        publicationDate="24 September 2018"
        link="url/to/article"
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('NewsItem onChange handler is triggered', () => {
    const onClickSpy = jest.fn();

    const component = shallow(
      <NewsItem
        headline="headline"
        onClick={onClickSpy}
        id="id"
        snippet="snippet"
        image="https://static01.nyt.com/images/2018/09/25/arts/25newamsterdam1/merlin_144237420_bf8f658f-7574-4665-b650-051b5cf0dec6-articleLarge.jpg"
        isActive={false}
        publicationDate="24 September 2018"
        link="https://www.nytimes.com/2018/09/24/arts/television/new-amsterdam-review-nbc.html"
      />
    );

    component.simulate('click');

    expect(onClickSpy).toHaveBeenCalled();
    expect(onClickSpy).toHaveBeenCalledWith('id');
    component.unmount();
  });

  it('NewsItem onChange handler is triggered', () => {
    const component = shallow(
      <NewsItem
        headline="headline"
        onClick={() => {}}
        id="id"
        snippet="snippet"
        image="https://static01.nyt.com/images/2018/09/25/arts/25newamsterdam1/merlin_144237420_bf8f658f-7574-4665-b650-051b5cf0dec6-articleLarge.jpg"
        isActive={true}
        publicationDate="24 September 2018"
        link="https://www.nytimes.com/2018/09/24/arts/television/new-amsterdam-review-nbc.html"
      />
    );

    component.setState = jest.fn();

    global.innerWidth = 500; // eslint-disable-line no-undef
    global.dispatchEvent(new Event('resize')); // eslint-disable-line no-undef

    return new Promise(resolve => {
      setTimeout(() => {
        expect(component.setState).toHaveBeenCalled();
        component.unmount();

        resolve();
      }, 400); // fait for debounced resize event handler will be executed
    });
  });
});
