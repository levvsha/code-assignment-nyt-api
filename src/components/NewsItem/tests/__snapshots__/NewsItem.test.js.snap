// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsItem /> NewsItem renders correctly 1`] = `
<div
  className="c-news-item"
  onClick={[Function]}
>
  <div
    className="item-media"
    style={null}
  >
    <div
      className="image-wrapper"
    >
      <img
        alt=""
        src="url/to/image.jpg"
      />
    </div>
  </div>
  <div
    className="item-body"
  >
    <h3
      className="item-title"
    >
      headline
    </h3>
    <div
      className="item-description"
    >
      <p
        className="item-published"
      >
        Publisherd: 
        24 September 2018
      </p>
      <p
        className="item-snippet"
      >
        snippet
      </p>
      <p
        className="read-more"
      >
        <a
          href="url/to/article"
          target="_blank"
        >
          Read more
        </a>
      </p>
    </div>
  </div>
</div>
`;
